### ISSUE TITLE
Gateway name missing after upgrading 1.2.5 to 1.3.8 or 1.3.9dev

### ISSUE THREAD

**Issue by mcary** [ingenesis/shopp#3253]
Orders created under 1.2.5 don't have a gateway visible after upgrading to 1.3.8 or 1.3.9dev.  Conversely, after downgrading from 1.3.x to 1.2.5, the orders created under 1.3.x do not have a visible gateway.  The space under the transaction ID is simply blank.

It seems the problem is that [core/ui/orders/orders.php:138](https://github.com/ingenesis/shopp/blob/1.3.x/core/ui/orders/orders.php#L138) does not have a $gateway because it's absent from $Gateways on [line 80](https://github.com/ingenesis/shopp/blob/1.3.x/core/ui/orders/orders.php#L80).

It looks like keys in that array changed from the form "OfflinePayment" to "ShoppOfflinePayment" or "PayPalStandard" to "ShoppPayPalStandard", but the data in the Orders table (purchase.gateway) is not upgraded, so the array lookups return null.

Is this a known issue?

**Comment by Just-m** [comment]
It can be solved in different ways. Not sure which one fits the Shopp flow best.

A workaround for now be to add a command at line 81 in the above mentioned file
`if ( empty($gateway) ) $gateway = $Gateways['Shopp'.$Order->gateway]->name;`

**Comment by mcary** [comment]
Yes, I have this in place already, but future upgrades to the plugin will 
remove the work-around.

Do you think it's worth adding a data migration step to the upgrade 
process?  Or maybe including a line like this into the core Shopp 
distribution until 1.4.x?

**Comment by Just-m** [comment]
I am not sure. I do not have a 1.2.5 install to test or check values with. In 1.3.8 the actual payment method label is stored in the database with every order. Might be better to just use that value instead of searching for the name/label in an active-gateway array. 

This approach would mean already present orders will use the old label in case you decide to change the label on an already active shop. Ah well an extra function on saving payment methods could take care of that :).

But... it is up to the lead programmer.

**Comment by mcary** [comment]
The "Charge Order" button does not show on the order page for orders placed before the upgrade to 1.3.x.  It looks like the cause is related.

In this case, the Gateway is looked up differently, by calling `$Purchase->gateway()`.  I think it could be fixed in a similar way:

```diff
         if ( isset($Gateways->active[ $processor ]) ) return $Gateways->active[ $processor ];
+        if ( isset($Gateways->active[ 'Shopp'.$processor ]) ) return $Gateways->active[ 'Shopp'.$processor ];
         else {
```

**Comment by mcary** [comment]
I see the pattern of looking up gateways by name elsewhere in the codebase as well, for example on the payments admin page.  I wonder if there is potential breakage in those areas?  I don't see any breakage, but maybe I'm just not familiar enough with how the app is intended to work or I don't have the right data in my installation to exercise it.  Perhaps it's worth an experienced Shopp developer's time to revisit that.

**Comment by jonathandavis** [comment]
Closed with merge of pull request #3256

**Comment by mcary** [comment]
Thanks!
